### PR TITLE
Add node delete controller

### PR DESCRIPTION
**Pull request by croomes**
Adds a controller to watch for Kubernetes Node delete events, and to remove the node from the StorageOS cluster if safe.

### CODE REVIEW

File: controllers/node-delete/controller.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package nodedelete
+
+import (
+	objectv1 "github.com/croomes/object-reconciler/controller/v1"
```
Comment (croomes) [inline controllers/node-delete/controller.go:4]: I'll move this package to storageos GH namespace.

Comment (croomes) [reply controllers/node-delete/controller.go]: done by removing the dependency, it wasn't adding much.

File: controllers/node-delete/controller.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	err := r.api.DeleteNode(req.Name)
+	if err != nil && err != storageos.ErrNamespaceNotFound {
+		// Re-queue without error.  We will get frequent transient errors, such
+		// as version conflicts or locked objects - that's ok.  Even refusal to
+		// delete due to the node still showing as online should be seen as
+		// transient and should not raise an error.
+		return ctrl.Result{Requeue: true}, nil
+	}
+	// We can't return an event as it requires an object, and we can't rely on
+	// the object still being present.
```
Comment (darkowlzz) [inline controllers/node-delete/controller.go:63]: Maybe this comment can be removed. We don't publish any event in this controller.

File: controllers/node-delete/controller.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+// Update determines whether an object update should trigger a reconcile.
+func (ChangePredicate) Update(event.UpdateEvent) bool {
+	return false
+}
+
+// Delete determines whether an object delete should trigger a reconcile.
+func (ChangePredicate) Delete(e event.DeleteEvent) bool {
+	// Ignore objects without metadata.
+	return e.Meta != nil
```
Comment (darkowlzz) [inline controllers/node-delete/controller.go:85]: We should also be able to check the node annotations to find out if storageos is installed on the node and ignore the events from nodes that don't have storageos.

File: controllers/node-delete/controller.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+// Reconcile deletes the StorageOS node.  The delete will fail if the control
+// plane has not yet determined that the node is offline.  Any errors will
+// result in a requeue, with standard back-off retries.
+//
+// Events are not sent as they require an object. By this point the namespace
+// object will no longer be available.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	err := r.api.DeleteNode(req.Name)
+	if err != nil && err != storageos.ErrNamespaceNotFound {
```
Comment (darkowlzz) [inline controllers/node-delete/controller.go:55]: Should this be node not found error?
I don't think nodes are namespaced.

File: controllers/node-delete/controller.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+// driver annotation.  The annotation is added by the drivers' node registrar,
+// so should only be present on nodes that have run StorageOS.
+func hasStorageOSDriverAnnotation(annotations map[string]string) bool {
+	drivers, ok := annotations[DriverAnnotationKey]
+	if !ok {
+		return false
+	}
+	driversMap := map[string]string{}
+	if err := json.Unmarshal([]byte(drivers), &driversMap); err != nil {
+		return false
```
Comment (darkowlzz) [inline controllers/node-delete/controller.go:110]: Logging this error would be helpful in debugging when it fails. Avoiding silent failures.

**Review by darkowlzz — APPROVED** [review]
LGTM!
A lot of nice tests. Thanks.

File: controllers/node-delete/controller.go (hunk 5 above)
Comment (croomes) [reply controllers/node-delete/controller.go]: Agreed, added

**Pull request merged**